sklearn/metrics/pairwise.py

```diff
@@ -436,6 +437,84 @@ def nan_euclidean_distances(X, Y=None, *, squared=False,
     return distances
 
 
+def _nan_fill_row_norm(r, fill_value: float = 0.):
+    p = r * r
+    p = np.nan_to_num(p, nan=fill_value)
+    return np.sum(p)
+
+
+@njit
+def _nan_fill_dot(X, Y, fill_values: np.ndarray):
+    if X.shape[1] != fill_values.shape[0]:
+        raise ValueError('X and fill_values have incompatible shapes')
+
+    p = np.zeros((X.shape[0], Y.shape[0],))
+    for i in range(0, X.shape[0]):
+        v1 = X[i, :]
+        for j in range(0, Y.shape[0]):
+            v2 = Y[j, :]
+            s = 0.
+            for k in range(0, fill_values.shape[0]):
+                if np.isnan(v1[k]) and np.isnan(v2[k]):
+                    pass
+                elif np.isnan(v1[k]) or np.isnan(v2[k]):
+                    s += fill_values[k]
+                else:
+                    s += v1[k] * v2[k]
+
+            p[i, j] = s
+
+    return p
+
+
+def nan_filled_euclidean_distances(X, fill_values, Y=None, squared=False, missing_values=np.nan, copy=True):
+    if X.dtype == np.float32 or (Y is not None and Y.dtype == np.float32):
+        raise NotImplementedError('X and Y must be np.float64 type')
+
+    X, Y = check_pairwise_arrays(X, Y, accept_sparse=False, force_all_finite='allow-nan', copy=copy)
+
+    # Get missing mask for X
+    missing_X = _get_mask(X, missing_values)
+
+    # Get missing mask for Y
+    missing_Y = missing_X if Y is X else _get_mask(Y, missing_values)
+
+    # calculate euclidean distances
+    XX = np.apply_along_axis(_nan_fill_row_norm, 1, X)
+    XX = XX[:, np.newaxis]
```

Comment:
what is this row for?

Reply:
It's just another way to reshape https://medium.com/@ian.dzindo01/what-is-numpy-newaxis-and-when-to-use-it-8cb61c7ed6ae